bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v2/PWAImpl.java

```diff
@@ -0,0 +1,137 @@
+/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ ~ Copyright 2020 Adobe
+ ~
+ ~ Licensed under the Apache License, Version 2.0 (the "License");
+ ~ you may not use this file except in compliance with the License.
+ ~ You may obtain a copy of the License at
+ ~
+ ~     http://www.apache.org/licenses/LICENSE-2.0
+ ~
+ ~ Unless required by applicable law or agreed to in writing, software
+ ~ distributed under the License is distributed on an "AS IS" BASIS,
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+
+package com.adobe.cq.wcm.core.components.internal.models.v2;
+
+import javax.annotation.Nonnull;
+import javax.annotation.PostConstruct;
+
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ValueMap;
+import org.apache.sling.models.annotations.Model;
+import org.apache.sling.models.annotations.injectorspecific.Self;
+
+import com.adobe.cq.wcm.core.components.models.PWA;
+import com.day.cq.wcm.api.Page;
+import com.day.cq.wcm.api.PageManager;
+
+@Model(adaptables = Resource.class,
+    adapters = {PWA.class})
+public class PWAImpl implements PWA {
+
+    static final String CONTENT_PATH = "/content/";
+
+    private boolean isEnabled = false;
+    private String manifestPath = "";
+    private String serviceWorkerPath = "";
+    private String themeColor = "";
+    private String iconPath = "";
+
+    @Self
+    private Resource resource;
+
+    @PostConstruct
+    protected void initModel() {
+        ValueMap valueMap = resource.getValueMap();
+        Boolean isPWAEnabled = valueMap.get(PN_ENABLE_PWA, Boolean.class);
+        this.isEnabled = (isPWAEnabled != null) ? isPWAEnabled : false;
+        if (!this.isEnabled) {
+            return;
+        }
+
+        this.themeColor = colorToHex(valueMap.get(PN_THEME_COLOR, ""));
+        this.iconPath = valueMap.get(PN_PWA_ICON, "");
+
+        String startURL = valueMap.get(PN_START_URL, "");
+        this.manifestPath = replaceSuffix(startURL, MANIFEST_NAME);
+
+        PageManager pageManager = resource.getResourceResolver().adaptTo(PageManager.class);
+        if (pageManager != null) {
+            Page page = pageManager.getContainingPage(resource);
+            if (page != null) {
+                String mappingName = page.getPath().replace("/", ".").substring(CONTENT_PATH.length());
+                this.serviceWorkerPath = "/" + mappingName + "sw.js";
```

Comment:
Make sure that `mappingName` is neither null nor empty

Reply:
mappingName will not be null or empty. I am already checking the page is not null. I have adjust to use replace rather than substring to further avoid any issue.